File: src/main/java/io/github/ytg1234/songsounds/base/song/Section.java
package io.github.ytg1234.songsounds.base.song;

import com.google.gson.*;
import io.github.ytg1234.songsounds.SongSounds;
import io.github.ytg1234.songsounds.util.ModVars;
import net.minecraft.util.JsonHelper;
import org.apache.logging.log4j.Level;

import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.util.Arrays;

public class Section {
    public String name;
    public float[] notes;

    private Section(String name, float... notes) {
        this.name = name;
        this.notes = notes;
    }

    public static class Serializer implements JsonDeserializer<Section> {
        @Override public Section deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            // Reading notes as String array
            JsonObject jsonObject = JsonHelper.asObject(json, "section");
            String name = JsonHelper.getString(jsonObject, "name");
            String[] notes = JsonHelper.deserialize(jsonObject, "notes", new String[0], context, String[].class);

            // Creating float array for future use
            float[] floatingNotes = new float[notes.length];

            // Populating the float array
            for (int i = 0; i < notes.length; i++) {
                try {
                    Field field = Notes.class.getField(notes[i].toUpperCase());
                    field.setAccessible(true);
                    floatingNotes[i] = (float) field.get(null);
                } catch (NoSuchFieldException | IllegalAccessException e) {
                    SongSounds.log(Level.WARN, "Datapack declared note " + notes[i] + ", which is unknown!");
                    floatingNotes[i] = ModVars.NoSuchNote;
                }
            }
            Section output = new Section(name, floatingNotes);
            SongSounds.log(Level.DEBUG, "Successfully deserialized a JSON section object, name is " + output.name + ", notes are " +
                    Arrays.toString(output.notes));
            return output;
        }
    }
}

File: src/main/kotlin/io/github/ytg1234/songsounds/base/song/Section.kt
package io.github.ytg1234.songsounds.base.song

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import io.github.ytg1234.songsounds.logger
import io.github.ytg1234.songsounds.util.NO_SUCH_NOTE
import net.minecraft.util.JsonHelper
import java.lang.reflect.Type

class Section private constructor(var name: String, vararg var notes: Float) {
    class Serializer : JsonDeserializer<Section> {
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Section {
            // Reading notes as String array
            val jsonObject = JsonHelper.asObject(json, "section")
            val name = JsonHelper.getString(jsonObject, "name")
            val notes = JsonHelper.deserialize(
                jsonObject, "notes", arrayOf(), context,
                Array<String>::class.java
            )

            // Creating float array for future use
            val floatingNotes = FloatArray(notes.size)

            // Populating the float array
            for (i in notes.indices) {
                val floatingNote = Notes[notes[i]]
                floatingNotes[i] = floatingNote
                if (floatingNote == NO_SUCH_NOTE) {
                    logger.warn("Datapack declared note ${notes[i]}, which is unknown!")
                }
            }
            val output = Section(name, *floatingNotes)
            logger.debug(
                "Successfully deserialized a JSON section object, name is ${output.name}, notes are ${output.notes.contentToString()}"
            )
            return output
        }
    }
}
